File: app/src/main/java/com/worldventures/dreamtrips/modules/tripsimages/model/Photo.java
package com.worldventures.dreamtrips.modules.tripsimages.model;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;

import com.esotericsoftware.kryo.DefaultSerializer;
import com.esotericsoftware.kryo.serializers.CompatibleFieldSerializer;
import com.worldventures.dreamtrips.core.ui.fragment.ImagePathHolder;
import com.worldventures.dreamtrips.modules.common.model.User;
import com.worldventures.dreamtrips.modules.common.view.custom.tagview.viewgroup.newio.model.PhotoTag;
import com.worldventures.dreamtrips.modules.feed.model.BaseFeedEntity;
import com.worldventures.dreamtrips.modules.trips.model.Location;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@DefaultSerializer(CompatibleFieldSerializer.class)
public class Photo extends BaseFeedEntity implements ImagePathHolder, Parcelable {

   private String title;
   private Date shotAt;
   private Date createdAt;
   private Location location;
   private List<String> tags;
   private String url;
   private List<PhotoTag> photoTags;
   private int photoTagsCount;
   private int width;
   private int height;

   public Photo() {
   }

   public Photo(String uid) {
      this.uid = uid;
   }

   @Override
   public String place() {
      return location != null ? location.getName() : null;
   }

   public Date getShotAt() {
      return shotAt;
   }

   public void setShotAt(Date shotAt) {
      this.shotAt = shotAt;
   }

   /*
    * Can be null if we get photo as attachment from feed
    * There are complications to add createdAt in feed on server
    */
   @Nullable
   public Date getCreatedAt() {
      return createdAt;
   }

   public void setCreatedAt(Date createdAt) {
      this.createdAt = createdAt;
   }

   public List<String> getTags() {
      if (tags == null) tags = new ArrayList<>();
      return tags;
   }

   public void setTags(List<String> tags) {
      this.tags = tags;
   }

   public Location getCoordinates() {
      return location;
   }

   public void setCoordinates(Location coordinates) {
      this.location = coordinates;
   }

   @NotNull
   public String getTitle() {
      return title != null ? title : "";
   }

   public void setTitle(String title) {
      this.title = title;
   }

   public void setUrl(String url) {
      this.url = url;
   }

   public List<PhotoTag> getPhotoTags() {
      if (photoTags == null) photoTags = new ArrayList<>();
      return photoTags;
   }

   public void setPhotoTags(List<PhotoTag> photoTags) {
      this.photoTags = photoTags;
   }

   public int getPhotoTagsCount() {
      return photoTagsCount;
   }

   public void setPhotoTagsCount(int photoTagsCount) {
      this.photoTagsCount = photoTagsCount;
   }

   public int getWidth() {
      return width;
   }

   public void setWidth(int width) {
      this.width = width;
   }

   public int getHeight() {
      return height;
   }

   public void setHeight(int height) {
      this.height = height;
   }

   @Override
   public String getOriginalText() {
      return getTitle();
   }

   @Override
   public String getImagePath() {
      return url;
   }

   public Location getLocation() {
      return location;
   }

   public void setLocation(Location location) {
      this.location = location;
   }

   ///////////////////////////////////////////////////////////////////////////
   // Parcelable
   ///////////////////////////////////////////////////////////////////////////

   public static final Creator<Photo> CREATOR = new Creator<Photo>() {
      @Override
      public Photo createFromParcel(Parcel in) {
         return new Photo(in);
      }

      @Override
      public Photo[] newArray(int size) {
         return new Photo[size];
      }
   };

   @Override
   public int describeContents() {
      return 0;
   }

   @Override
   public void writeToParcel(Parcel parcel, int i) {
      parcel.writeString(uid);
      parcel.writeInt(commentsCount);
      parcel.writeInt(likesCount);
      parcel.writeInt(liked ? 1 : 0);
      parcel.writeString(title);
      parcel.writeSerializable(shotAt);
      parcel.writeSerializable(createdAt);
      parcel.writeParcelable(location, i);
      parcel.writeStringList(tags);
      parcel.writeString(url);
      parcel.writeParcelable(owner, i);
      parcel.writeTypedList(photoTags);
      parcel.writeInt(photoTagsCount);
      parcel.writeInt(width);
      parcel.writeInt(height);
   }

   protected Photo(Parcel in) {
      uid = in.readString();
      commentsCount = in.readInt();
      likesCount = in.readInt();
      liked = in.readInt() == 1;
      title = in.readString();
      shotAt = (Date) in.readSerializable();
      createdAt = (Date) in.readSerializable();
      location = in.readParcelable(Location.class.getClassLoader());
      tags = in.createStringArrayList();
      url = in.readString();
      owner = in.readParcelable(User.class.getClassLoader());
      photoTags = new ArrayList<>();
      in.readTypedList(photoTags, PhotoTag.CREATOR);
      photoTagsCount = in.readInt();
      width = in.readInt();
      height = in.readInt();
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/social/ui/tripsimages/model/Photo.kt
package com.worldventures.dreamtrips.social.ui.tripsimages.model

import android.os.Parcel
import android.os.Parcelable
import com.esotericsoftware.kryo.DefaultSerializer
import com.esotericsoftware.kryo.serializers.CompatibleFieldSerializer
import com.worldventures.core.model.ImagePathHolder
import com.worldventures.core.model.User
import com.worldventures.dreamtrips.modules.trips.model.Location
import com.worldventures.dreamtrips.social.ui.feed.model.BaseFeedEntity
import com.worldventures.dreamtrips.social.ui.feed.model.comment.Comment
import com.worldventures.dreamtrips.social.ui.feed.view.custom.tagview.viewgroup.newio.model.PhotoTag
import java.util.Date

@DefaultSerializer(CompatibleFieldSerializer::class)
class Photo() : BaseFeedEntity(), ImagePathHolder, Parcelable {

   var title: String? = null
   var shotAt: Date? = null
   /*
    * Can be null if we get photo as attachment from feed
    * There are complications to add createdAt in feed on server
   */
   // Workaround property as backing field to createdAt, refactor after we migrate BaseFeedEntity to Kotlin
   private var photoCreatedAt: Date? = null
   var location: Location? = null
   var tags: MutableList<String> = mutableListOf()
   var url: String = ""
   var photoTags: MutableList<PhotoTag> = mutableListOf()
   var photoTagsCount: Int = 0
   var width: Int = 0
   var height: Int = 0

   constructor(uid: String) : this() {
      this.uid = uid
   }

   override fun place(): String? {
      return if (location != null) location?.let { it.name } else null
   }

   override fun getOriginalText(): String? = title

   override fun getImagePath(): String? = url

   override fun getCreatedAt(): Date? = photoCreatedAt

   fun setCreatedAt(value: Date?) {
      this.photoCreatedAt = value
   }

   constructor(source: Parcel) : this() {
      this.uid = source.readString()
      this.owner = source.readParcelable(User::class.java.classLoader)
      this.commentsCount = source.readInt()
      this.comments = source.createTypedArrayList(Comment.CREATOR)
      this.liked = source.readInt() == 1
      this.likesCount = source.readInt()
      this.language = source.readString()
      this.firstLikerName = source.readString()
      this.translation = source.readString()
      this.isTranslated = source.readInt() == 1

      this.title = source.readString()
      this.shotAt = source.readSerializable() as Date?
      this.photoCreatedAt = source.readSerializable() as Date?
      this.location = source.readSerializable() as Location?
      this.tags = source.createStringArrayList()
      this.url = source.readString()
      this.photoTags = source.createTypedArrayList(PhotoTag.CREATOR)
      this.photoTagsCount = source.readInt()
      this.width = source.readInt()
      this.height = source.readInt()
   }

   override fun describeContents() = 0

   override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
      writeString(uid)
      writeParcelable(owner, 0)
      writeInt(commentsCount)
      writeTypedList(comments)
      writeInt(if (liked) 1 else 0)
      writeInt(likesCount)
      writeString(language)
      writeString(firstLikerName)
      writeString(translation)
      writeInt(if (isTranslated) 1 else 0)

      writeString(title)
      writeSerializable(shotAt)
      writeSerializable(photoCreatedAt)
      writeSerializable(location)
      writeStringList(tags)
      writeString(url)
      writeTypedList(photoTags)
      writeInt(photoTagsCount)
      writeInt(width)
      writeInt(height)
   }

   companion object {
      @JvmField
      val CREATOR: Parcelable.Creator<Photo> = object : Parcelable.Creator<Photo> {
         override fun createFromParcel(source: Parcel): Photo = Photo(source)
         override fun newArray(size: Int): Array<Photo?> = arrayOfNulls(size)
      }
   }
}
